runtime/strategies/match-recipe-by-verb.js

```diff
@@ -27,20 +38,80 @@ export default class MatchRecipeByVerb extends Strategy {
         if (particle.allConnections().length > 0)
           return;
 
-        if (Object.keys(particle.consumedSlotConnections).length > 0)
-          return;
-
         let recipes = arc.context.findRecipesByVerb(particle.primaryVerb);
 
+        let slotConstraints = {};
+        for (let consumeSlot of Object.values(particle.consumedSlotConnections)) {
+          let targetSlot = consumeSlot.targetSlot.sourceConnection ? consumeSlot.targetSlot : null;
+          slotConstraints[consumeSlot.name] = {targetSlot, providedSlots: {}};
+          for (let providedSlot of Object.keys(consumeSlot.providedSlots)) {
+            let sourceSlot = consumeSlot.providedSlots[providedSlot].consumeConnections.length > 0 ? consumeSlot.providedSlots[providedSlot] : null;
+            slotConstraints[consumeSlot.name].providedSlots[providedSlot] = sourceSlot;
+          }
+        }
+
+        recipes = recipes.filter(recipe => MatchRecipeByVerb.satisfiesSlotConstraints(recipe, slotConstraints));
+
         return recipes.map(recipe => {
           return (outputRecipe, particle) => {
-            recipe.mergeInto(outputRecipe);
+            let {handles, particles, slots} = recipe.mergeInto(outputRecipe);
+
             particle.remove();
 
+            for (let consumeSlot in slotConstraints) {
+              if (slotConstraints[consumeSlot].targetSlot) {
+                let {mappedSlot} = outputRecipe.updateToClone({mappedSlot: slotConstraints[consumeSlot].targetSlot});
+                for (let particle of particles) {
+                  if (particle.consumedSlotConnections[consumeSlot] && particle.consumedSlotConnections[consumeSlot].targetSlot == null) {
+                    particle.consumedSlotConnections[consumeSlot]._targetSlot = mappedSlot;
+                    mappedSlot._consumerConnections.push(particle.consumedSlotConnections[consumeSlot]); 
+                    break;
+                  }
+                }
+              }
+
+              for (let provideSlot in slotConstraints[consumeSlot].providedSlots) {
+                let slot = slotConstraints[consumeSlot].providedSlots[provideSlot];
+                let {mappedSlot} = outputRecipe.updateToClone({mappedSlot: slot});
+                for (let particle of particles) {
+                  if (particle.consumedSlotConnections[consumeSlot]) {
+                    if (particle.consumedSlotConnections[consumeSlot].providedSlots[provideSlot]) {
+                      let oldSlot = particle.consumedSlotConnections[consumeSlot].providedSlots[provideSlot];
+                      oldSlot.remove();
+                      particle.consumedSlotConnections[consumeSlot].providedSlots[provideSlot] = mappedSlot;
```

Comment:
This looks safe, maybe 🤷‍♂️ , but I think the API intended that we treat the `providedSlots` object as immutable.